### PR TITLE
Fix wrong imports

### LINKED ISSUE #97: Factory generation: inject nested class into nested class
Anvil: 2.0,2

File1:
```
package com.example.myapplication

import com.example.myapplication.foo.Foo
import javax.inject.Inject

class MyClass (bar: Bar): Foo(bar) {
  class MyClassInner @Inject constructor(bar: Bar)
}
```

File2
```
package com.example.myapplication.foo

import javax.inject.Inject

abstract class Foo constructor(bar: Bar) {
    class Bar @Inject constructor()
}
```

Generated `MyClass_MyClassInner_Factory` has incorrect imports

### PR DESCRIPTION
**Pull request by vRallev**
This PR reworks how imports are resolved. Before I implemented an "easy fix, "workaround" or "hack" 🙃  As usual, those things bite one in the long run. Now I resolve all types properly so that Kotlin Poet can add the correct imports.

Fixes #97, #82

### CODE REVIEW

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/ProvidesMethodFactoryGenerator.kt [hunk 1]
```diff
@@ -83,18 +83,12 @@ internal class ProvidesMethodFactoryGenerator : CodeGenerator {
 
     val parameters = function.valueParameters.mapToParameter(module)
 
-    // This is a little hacky. The typeElement could be "String", "Abc", etc., but also a generic
-    // type like "List<String>". We simply copy this literal as return type into our generated
-    // code. Kotlinpoet will add an import for this literal like "import String", which we later
-    // will remove again.
-    //
-    // This solution is a lot easier than trying to resolve all FqNames for each type.
```
Comment (vRallev) [inline compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/ProvidesMethodFactoryGenerator.kt:91]: That was the hack.

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/KotlinPoet.kt [hunk 2]
```diff
@@ -204,58 +268,3 @@ internal fun FileSpec.writeToString(): String {
   }
   return stream.toString()
 }
-
-/**
- * Removes invalid imports that can be caused by [requireTypeName] and copies imports from the
- * source file. Copying imports is necessary for star imports.
- */
-internal fun String.replaceImports(clazz: KtClassOrObject): String {
```
Comment (vRallev) [inline compiler/src/main/java/com/squareup/anvil/compiler/codegen/KotlinPoet.kt:212]: That workaround isn't need anymore. The test coverage is really high, that's why I'm confident in this new solution.

**Comment by vRallev** [comment]
I was a little afraid of the overhead compared to the old solution, but it is negligible. I ran benchmarks over the weekend and the overhead is around 1% (I wonder if this actually came from somewhere else on the machine).

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/KotlinPoet.kt [hunk 3]
```diff
@@ -9,61 +9,131 @@ import com.squareup.anvil.compiler.providerFqName
...
-  return try {
-    ClassName.bestGuess(asString())
-  } catch (e: IllegalArgumentException) {
-    // This happens when the class name starts with a lowercase character.
-    TypeVariableName(asString())
+private fun FqName.asClassName(module: ModuleDescriptor): ClassName {
+  try {
+    return ClassName.bestGuess(asString())
+  } catch (ignored: IllegalArgumentException) {
+    // Probably lowercase class.
```
Comment (steve-the-edwards) [inline compiler/src/main/java/com/squareup/anvil/compiler/codegen/KotlinPoet.kt:65]: Nit: could comment intentional fall-through.

File: compiler/src/test/java/com/squareup/anvil/compiler/dagger/TestClasses.kt [hunk 4]
```diff
@@ -0,0 +1,11 @@
+@file:Suppress("unused")
+
+package com.squareup.anvil.compiler.dagger
+
+// These classes is used in some of the unit tests. Don't touch them.
```
Comment (steve-the-edwards) [inline compiler/src/test/java/com/squareup/anvil/compiler/dagger/TestClasses.kt:5]: nit: "... are used ..."

**Review by steve-the-edwards — APPROVED** [review]

**Pull request merged**